### PR TITLE
feat(ton): add TON network support (native + jetton)

### PR DESCRIPTION
**Pull request by web3luka**
## Description

This PR adds **TON (The Open Network) support** to x402.  
It introduces a new network adapter that enables merchants and facilitators to accept payments in TON native coins as well as Jettons ([TEP-74 standard](https://github.com/ton-blockchain/TEPs/blob/master/text/0074-jettons-standard.md)).  

**Key changes:**
- [x] Implemented `verifyTONExact` for native and jetton payments  
- [x] Added address normalization across raw/bounceable / non-bounceable formats  
- [x] Added memo validation (`x402:` strict mode) with fallback for legacy memos  
- [x] Unified error codes for better integration with existing adapters  
- [x] Utility helpers: `normalizeTonAddress`, `getTonExplorerUrl`, `validateMemo`, `toAtomic`  
- [x] Unit tests covering native + jetton flows, decimal mismatch, forward_ton_amount edge cases, cross-format addresses  
- [x] Explorer integration via tonviewer.com (mainnet + testnet)  

---

## Tests

- [x] Added `src/ton/index.test.ts` with >180 cases:
  - Native transfers with memo  
  - Jetton transfers with forward payload (memo)  
  - Decimal mismatch/amount mismatch  
  - Cross-format address normalization  
  - Invalid memo and expired invoice handling  
- [x] Verified unit tests with `pnpm test` inside `/typescript`  
- [x] Manually checked explorer URL generation for mainnet and testnet  

---

## Checklist

- [x] I have formatted and linted my code (`pnpm format && pnpm lint`)  
- [x] All new and existing tests pass (`pnpm test`)  
image
- [x] My commits are signed (verified via `git log --show-signature`)

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: typescript/packages/x402/src/ton/jettonPayload.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+export function encodeJettonTransfer(p: BuildJettonMsg) {
+  // Build the Jetton transfer body using ton-core (TEP-74 compliant)
+  // op::transfer#0f8a7ea5 query_id:uint64 amount:(VarUInteger 16) destination:MsgAddress
+  // response_destination:MsgAddress custom_payload:(Maybe ^Cell) forward_ton_amount:(VarUInteger 16)
+  // forward_payload:(Either Cell ^Cell) = InternalMsgBody;
+  const transferBody = beginCell()
+    .storeUint(0x0f8a7ea5, 32) // op code for transfer (TEP-74)
+    .storeCoins(BigInt(p.amount)) // jetton amount (atomic units)
+    .storeAddress(Address.parse(p.to)) // destination address
+    .storeAddress(null) // response_destination (null = burn remaining jettons)
```
Comment (ilyaqznetsow) [inline typescript/packages/x402/src/ton/jettonPayload.ts]: why burn change for sending jettons? its better to return to the sender, because blockchain fees is a subject of change

**Comment by phdargen** [comment]
Thanks a lot for your contribution @web3luka!
Now that v2 is released, the legacy packages are effectively frozen and we are closing out PRs targeting v1.
Please feel free to open a new PR against the v2 package, we'd be happy to review it. 
Here are some resources to get started with v2:

Migration Guide: https://x402.gitbook.io/x402/guides/migration-v1-to-v2
Contributing Guide: https://github.com/coinbase/x402/blob/main/CONTRIBUTING.md

**Comment by phdargen** [comment]
Please have a look at the contributing guide on how to add new chains here: https://github.com/coinbase/x402/blob/main/CONTRIBUTING.md#new-chains. The first step would be to write a document that outlines the implementation for the exact payment scheme on your chain, see eg for EVM: https://github.com/coinbase/x402/blob/main/specs/schemes/exact/scheme_exact_evm.md. Open a PR with that document and the team will review it and provide feedback. Please do not jump immediately to the actual implementation, just the document is enough to start the process
